src/utils/fileInspector.js

```diff
@@ -0,0 +1,68 @@
+import { Gunzip, strFromU8 } from 'fflate';
+
+import techOptions from './fileUploadSpecifications';
+import readFileToBuffer from './readFileToBuffer';
+
+const Verdict = {
+  INVALID_NAME: -2,
+  INVALID_FORMAT: -1,
+  VALID_ZIPPED: 0,
+  VALID_UNZIPPED: 1,
+};
+
+const MATRIX_SIGNATURE = Buffer.from('%%MatrixMarket', 'ascii');
+const FEATURES_SIGNATURE = Buffer.from('ENS', 'ascii');
+
+const GZIP_SIGNATURE = Buffer.from([0x1f, 0x8b]);
+
+const inspectFile = async (file, technology) => {
+  // Validate a file requested for upload to the platform.
+
+  // immediately discard file if filename is not in valid set
+  const validNames = techOptions[technology].acceptedFiles;
+  if (!validNames.has(file.name)) {
+    return Verdict.INVALID_NAME;
+  }
+
+  // if name is valid, inspect first 16 bytes to validate format
+  let data = await readFileToBuffer(file.slice(0, 16));
+
+  const isGzipped = !data.slice(0, 2).compare(GZIP_SIGNATURE);
+
+  if (isGzipped) {
+    // if gzipped, decompress a small chunk to further validate contents
+    // eslint-disable-next-line no-unused-vars
+    const gunzip = new Gunzip((chunk, _) => {
+      data = Buffer.from(chunk.slice(0, 16));
+    });
+    gunzip.push(await readFileToBuffer(file.slice(0, 128)));
```

Comment:
This is more convoluted than it ought to be but it was the only way I could get fflate to decompress only a portion of the file - using `gunzipSync` for example with a slice would just hang